File: Array/llb/169. Majority Element.cpp
/*
问题描述：
Given an array of size n, find the majority element. The majority element is the element that appears more than ⌊ n/2 ⌋ times.

You may assume that the array is non-empty and the majority element always exist in the array.

Credits:
Special thanks to @ts for adding this problem and creating all test cases.
*/




class Solution {
public:
    int majorityElement(vector<int>& nums) {
        int majorityIndex = 0;
        int count = 1;        //初始化为第一个
        for (int i=1; i<nums.size(); i++) {
            if (nums[i] == nums[majorityIndex]) {
                count++;
            }
            else {
                count--;
            }
            if (count == 0) {
                majorityIndex = i;
                count = 1;
            }
        }
        
        return nums[majorityIndex];
    }
};

File: BitManipulation/169_MajorityElement.py
# coding: utf8


"""
    题目链接: https://leetcode.com/problems/majority-element/description.
    题目描述:

    Given an array of size n, find the majority element. The majority element is the element that appears more than
    ⌊ n/2 ⌋ times.

    You may assume that the array is non-empty and the majority element always exist in the array.

    Credits:
    Special thanks to @ts for adding this problem and creating all test cases.

"""


class Solution(object):
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        ans = nums[0]
        cnt = 1

        for i in range(1, len(nums)):
            if cnt == 0:
                cnt += 1
                ans = nums[i]
            elif nums[i] == ans:
                cnt += 1
            else:
                cnt -= 1

        return ans
